### DISCUSSION TITLE
Is it possible to read environment variables from yaml configuration files?

### DISCUSSION THREAD

**Discussion by abbluiz** [apache/apisix#5222] [Q&A]
I want to use different configuration depending on my environment on some YAML configuration files, specially in `conf/apisix.yaml` in stand-alone mode. Is it possible to read from environment variable like `${VAR}` inside YAML file?

**Comment by tzssangglass** [comment]
refer to:https://github.com/apache/apisix/blob/master/docs/en/latest/profile.md

**Reply by abbluiz** [reply]
I saw this, however it is not flexible enough for me. I would like do store OAuth 2.0 endpoints and client secret in environment variables and use them inside the files like `${OAUTH_INTROSPECTION_ENDPOINT}`. This is because I commit my route config whilst using stand alone mode.

**Reply by tzssangglass** [reply]
try use `${{VAR}}` in `config.yaml`, refer to https://github.com/apache/apisix/blob/a413014debc00b79566a8607ca4a340bd76d9068/conf/config.yaml#L24

**Reply by abbluiz** [reply]
Yes, but what about other configuration files, like `apisix.yaml` when using stand alone mode? It seems that this syntax does not work for it.

**Comment by tzssangglass** [comment]
sounds like a good feature.

**Reply by tzssangglass** [reply]
@spacewander @tokers @membphis what's your options?

**Reply by abbluiz** [reply]
Yes, it would be great to be natively supported. To anyone interested, for now I am creating a temporary docker container that parses the yaml file and replaces with my environment variables using a Python script.

**Reply by tokers** [reply]
I think so, it's my understanding that secrets are not suitable to be written in the configuration directly, env should be supported.

**Reply by spacewander** [reply]
The variable resolve is done in https://github.com/apache/apisix/blob/8ac00e0610eded0f8e2738c5fb6afc0776434552/apisix/cli/file.lua#L218

We can do the same thing with https://github.com/apache/apisix/blob/8ac00e0610eded0f8e2738c5fb6afc0776434552/apisix/core/config_yaml.lua#L60 after refactor.

PR is welcome!

**Reply by tzssangglass** [reply]
~~I've turned this discussion into an issue. refer to:https://github.com/apache/apisix/issues/5243~~ see:https://github.com/apache/apisix/issues/5244
